Feature/#10 webapp

oli-ver opened the pull request:
Here is the pull request for the web app feature #10.
I think we can still work on some of the details, but in general it's working.

oli-ver committed:
midi to part mp3s - preparation and download #10

Added documentation:
- how to build and start the docker container
- how to develop the web app
- how to open the webapp in the browser
- Pencil file with web app prototype
- Added formatter documentation
- Also added some missing fixtures docs

Changes in midi_to_part_mp3s:

- Add possibility to specify the output directory
- It's now possible to include an instrument with ID 0 (wrong if statement)
- Removed deleted common_solo_tracks parameter from add_accompaniment

Web App:

- Added flask app that returns a base bootstrap hello world template on the path /
- Added download feature at download/ that calls midi-to-part-mp3, creates a zip file that is returned to the user
- Implemented visualization of midi and musicxml after uploading it
- Added drag and drop features to select tracks from MusicXML or Midi and download the converted mp3s as zip
- webapp content resides in a sub folder
- The user can choose from some instruments
- UI disables buttons on clicks and enables again after request is finished

jackkinsella commented:
Nice one! 

This one will take me a while to review because I'd like to poke around with it in the browser.

oli-ver commented:
Sure, take your time.
I saw the branch had conflicts with the master so I rebased it and also squashed all the commits into one.

jackkinsella commented on webapp/app.py:
```diff
@@ -0,0 +1,152 @@
+
+
+@app.route("/download", methods=['POST'])
+def midi_to_mp3():
```
I know this is only a version 1 so take this as a request for a subsequent PR:  These methods are far too long and should be split up. For example (see next comment)

jackkinsella commented on webapp/app.py:
```diff
@@ -0,0 +1,152 @@
+@app.route("/download", methods=['POST'])
+def midi_to_mp3():
+    random_folder_name = _check_tmp_folder()
+    soprano_tracks = request.form.getlist(SOPRANO)
```
There could be a call to a method for converting the `request.form` input into an args object that would save a good few lines.

jackkinsella commented on webapp/app.py:
```diff
@@ -0,0 +1,152 @@
+
+    zip_file_name = '%s/%s' % (random_folder_name, FILENAME)
+    print ('Creating zip file: %s' % zip_file_name)
+    with ZipFile(zip_file_name, 'w') as myzip:
```
Maybe all this could be encapsulated in `send_zip_file`?

oli-ver replied:
Ok, I will change that.

jackkinsella commented on webapp/app.py:
```diff
@@ -0,0 +1,152 @@
+            music_xml)
+    ))
+
+    argument_list: List[str] = [midi_file_path]
```
The next few lines are really all about generating the track listings, so they could also be encapsulated into another method.

oli-ver replied:
Ok, on it.

jackkinsella commented on webapp/static/main.js:
```diff
@@ -0,0 +1,128 @@
+                // Create a new Blob object using the 
+                //response data of the onload object
+                var blob = new Blob([this.response], { type: 'application/zip' });
+                //Create a link element, hide it, direct 
```
In general, instead of some of these comments you could instead write short method that encapsulates the intent.

For example:

```javascript
function createHiddenLink() {
}
```

oli-ver replied:
I encapsulate it using a createBlobAndTriggerDownload() method.

jackkinsella commented on webapp/static/main.js:
```diff
@@ -0,0 +1,128 @@
+    ev.preventDefault();
+}
+
+function drag(ev) {
```
I don't know that `ev` is. Could you replace with a longer variable name here and in the other method signatures?

oli-ver replied:
It's an event object. I changed the name and removed drag() as it was not in use.

jackkinsella commented on webapp/static/main.js:
```diff
@@ -0,0 +1,128 @@
+        downloadFormData.append('musicxml', window.musicXml)
+        let xhr = new XMLHttpRequest();
+        //set the request type to post and the destination url to '/convert'
+        xhr.open('POST', url);
```
It seems odd to use the jQuery requesting helpers earlier in this file (line 14: `$.ajax()`), but now to switch to the built-in Javascript `xhr` stuff. Is this necessary? Could you choose one method and keep it consistent?

oli-ver replied:
Sure, I will use jQuery in both ajax requests.

jackkinsella reviewed (changes requested):
Everything worked when I tested the web-app manually. However the integration test failed (see comment). Getting the integration test to work is the only mandatory change before I will merge. Consider everything else optional recommendations / ideas for future refactoring.

Some (optional) improvements might be:
* some front-end indication work is in progress and it hasn't crashed (can take 20 seconds+)
* removal of undefined nesting in drag-drop
<img width="181" alt="Screen Shot 2019-06-29 at 16 47 19" src="https://user-images.githubusercontent.com/151206/60385788-c9e7f900-9a8d-11e9-8657-52f5f4079a03.png">
* removal or these weird instrument strings 
<img width="489" alt="Screen Shot 2019-06-29 at 16 43 35" src="https://user-images.githubusercontent.com/151206/60385793-d9674200-9a8d-11e9-82f9-0aafc5e2ae88.png">

oli-ver commented on README.md:
```diff
@@ -74,7 +74,7 @@ assume you name this `mpt_venv`.
 $ python3 -m venv mtp_venv
 $ source mpt_venv/bin/activate
-$ pip3 install -r requirements.txt
+$ pip install -r requirements.txt
```
With `python3 -m venv mtp_venv` you switch into the virtual environment. In this environment the commands `python3` and `pip3` do not exist (at least in my installation) and you have to use `pip` or `python`. Is this different for you?

oli-ver commented on webapp/app.py:
```diff
@@ -0,0 +1,152 @@
+from flask import Flask, render_template, request, send_file, Response
+from werkzeug import FileStorage
+from zipfile import ZipFile
```
I think ordering can be done automatically by the formatters in the IDEs. I suppose they implement [PEP-0008](https://www.python.org/dev/peps/pep-0008/#imports). We can check that later and create a little branch to fix this for all files.

oli-ver commented on webapp/app.py:
```diff
@@ -0,0 +1,152 @@
+
+app = Flask(__name__)
+
+FILE: str = 'file'
```
Ah right, we can move that to a class later.

oli-ver committed:
Feature #10 - Webapp review tasks

Implemented first review tasks:
* Refactored some long methods
* Used better variable names for HTML events
* Instead of XMLHttpRequest $.ajax is used in both occurences
Additional:
* Added error message to html form
* Added .dockerignore to not copy vscode config and pyhton venv into the docker container
* Changed order of commands in docker file to make better use of the caching (install python dependencies and then app)

oli-ver commented:
I worked on some of the review tasks but I am not completely done yet. After changing form XMLHttpRequest to jquery.ajax the download seems not to work anymore. The browser shows HTTP 200 but the flask endpoint seems not to be called. I will check that again, perhaps it's just because I used a different development environment.

I hope I can have a look at it tomorrow.

oli-ver committed:
Feature jackkinsella#10 - Webapp bug fixes

Fixed wrong argument list (probably not thought of that during rebase)
Fixed $.ajax for the file download to use the correct methods and set the blob data type directly in xhrFields because jQuery.ajax() does not support it in dataType

oli-ver committed:
Feature jackkinsella#10 - Documentation fix

Removed curl commando as the webapp in the meantime does not work with a file anymore but with musicxml

oli-ver commented:
Ok it works now again. I seem to have broken it during the rebase process and probably tested with the wrong version of the docker container (I cannot imagine how else I did not see that the -f parameter was missing)

The curl commando cannot be used for testing. During the development I changed the app. It's not using a file but musicxml provided by the user interface, so download/ cannot be called like that.
If we want have a webapi that supports a midi file directly we can add that later.

Can you please retest?

oli-ver committed:
Feature jackkinsella#10 - Acknowledgements

Added acknowledgements for the opensource libraries to the index.html

oli-ver commented:
As discussed in #10 this is out of scope for the upstream project.
